File: py_publisher_subscriber/src/py_publisher_subscriber/py_publisher_subscriber/subscriber_member_function.py
#! /usr/bin/env python3

"""
Description:
This ROS2 nose subscribes to "Hello, World!" messages.
---
Publishes to: None
---
Subscribes to:
The channel containing the "Hello, World!" messages
/my_topic (std_msgs/msg/String)
---
Author: Sai C Akella
Date: 18/05/2026
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from rclpy.executors import ExternalShutdownException


class MinimalSubscriber(Node):
    def __init__(self):
        super().__init__("minimal_subscriber")

        self._subscription = self.create_subscription(String,
                                                      "my_topic",
                                                      # callback gets called as soon as it receives
                                                      # a message.
                                                      self._listener_callback,
                                                      10
                                                      )

    def _listener_callback(self, msg):
        # simply prints an info message to the console, along with the data it received
        self.get_logger().info(f'I heard: "{msg.data}"')


def main(args=None):
    try:
        rclpy.init(args=args)
        node = MinimalSubscriber()
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        print()
    except Exception as e:
        print(f"An error occurred: {e}")
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

File: unit_tests/src/unit_tests/src/subscriber_member_function.cpp
/**
 * @file subscriber_member_function.cpp
 * @author Sai C Akella
 * @brief This ROS2 node subscribes to "Hello, World!" messages
 * @version 0.1
 * @date 2026-05-22
 *
 * ----------
 * Subscribes to:
 * /my_topic - std_msgs/msg/String
 *
 * Publishes to:
 * None
 */

#include <functional>
#include <memory>
#include "rclcpp/rclcpp.hpp" // ROS 2 C++ Client Lib
#include "std_msgs/msg/string.hpp" // Handling String messages

using std::placeholders::_1; //Placeholder for callback function

class MinimalSubscriber : public rclcpp::Node {
  public:
    MinimalSubscriber() : Node("minimal_subscriber"){
      subscription_ = this->create_subscription<std_msgs::msg::String>(
        "my_topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
    }

  private:
    void topic_callback(const std_msgs::msg::String & msg) const {
      RCLCPP_INFO(this->get_logger(), "I heard: %s", msg.data.c_str());
    }
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char* argv[]){
  rclcpp::init(argc, argv);
  auto node = std::make_shared<MinimalSubscriber>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
